File: app/src/main/java/com/hiddenramblings/tagmo/TagMo.java
package com.hiddenramblings.tagmo;

import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.text.Html;
import android.text.Spanned;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatDelegate;

import com.github.anrwatchdog.ANRError;
import com.github.anrwatchdog.ANRWatchDog;
import com.hiddenramblings.tagmo.browser.Preferences;
import com.hiddenramblings.tagmo.eightbit.content.ScaledContext;
import com.hiddenramblings.tagmo.eightbit.io.Debug;

import org.lsposed.hiddenapibypass.HiddenApiBypass;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.ref.SoftReference;
import java.util.Objects;

import me.weishu.reflection.Reflection;

public class TagMo extends Application {

    private static SoftReference<Context> mContext;
    public static final int uiDelay = 50;

    public static Context getContext() {
        return mContext.get();
    }

    private final boolean isWatchingANR = !BuildConfig.DEBUG && !BuildConfig.GOOGLE_PLAY;
    private boolean isUncaughtANR(Throwable error) {
        return null != error.getCause() && (error.getCause().getCause() instanceof ANRError);
    }

    public void setThemePreference() {
        switch (new Preferences(getApplicationContext()).applicationTheme()) {
            case 0:
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
                break;
            case 1:
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                break;
            case 2:
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                break;
        }
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        if (BuildConfig.WEAR_OS) {
            mContext = new SoftReference<>(new ScaledContext(this).watch(2));
            mContext.get().setTheme(R.style.AppTheme);
        } else {
            mContext = new SoftReference<>(this);
        }
        if (Debug.INSTANCE.isNewer(Build.VERSION_CODES.P))
            HiddenApiBypass.addHiddenApiExemptions("LBluetooth");
        else if (Debug.INSTANCE.isNewer(Build.VERSION_CODES.LOLLIPOP))
            Reflection.unseal(base);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
        if (BuildConfig.WEAR_OS) mContext.get().setTheme(R.style.AppTheme);

        if (isWatchingANR) {
            new ANRWatchDog(30000).setANRListener(error -> {
                StringWriter exception = new StringWriter();
                error.printStackTrace(new PrintWriter(exception));
                Debug.INSTANCE.processException(this, exception.toString());
            }).start();
        }

        Thread.setDefaultUncaughtExceptionHandler((t, error) -> {
            if (isWatchingANR && isUncaughtANR(error)) return;
            StringWriter exception = new StringWriter();
            error.printStackTrace(new PrintWriter(exception));
            Toast.makeText(this, R.string.logcat_crash, Toast.LENGTH_SHORT).show();
            Debug.INSTANCE.processException(this, exception.toString());
            System.exit(0);
        });

        setThemePreference();
    }

    private static final String commitHash = "#" + BuildConfig.COMMIT;
    private static final String versionLabel = "TagMo "
            + BuildConfig.VERSION_NAME + " (" + (BuildConfig.GOOGLE_PLAY
            ? "Google Play" : "GitHub") + " " + (BuildConfig.WEAR_OS
            ? "Wear OS" : Objects.equals(BuildConfig.BUILD_TYPE, "release")
            ? "Release" : "Debug") + ") " + commitHash;
    private static final String commitLink =
            "<a href=https://github.com/HiddenRamblings/TagMo/commit/"
            + BuildConfig.COMMIT + ">" + commitHash + "</a>";
    public static Spanned getVersionLabel(boolean plain) {
        return Html.fromHtml(plain ? versionLabel : versionLabel.replace(commitHash, commitLink));
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/TagMo.kt
package com.hiddenramblings.tagmo

import android.app.Application
import android.app.UiModeManager
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Parcelable
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.text.HtmlCompat
import com.hiddenramblings.tagmo.eightbit.content.ScaledContext
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.eightbit.os.Storage
import com.hiddenramblings.tagmo.eightbit.os.Version
import com.hiddenramblings.tagmo.eightbit.webkit.ChromeIntegration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import me.weishu.reflection.Reflection
import java.io.PrintWriter
import java.io.Serializable
import java.io.StringWriter
import kotlin.system.exitProcess

inline fun <reified T : Parcelable> Intent.parcelable(key: String): T? = when {
    Version.isTiramisu -> getParcelableExtra(key, T::class.java)
    else -> @Suppress("deprecation") getParcelableExtra(key) as? T
}
inline fun <reified T : Parcelable> Intent.parcelableArrayList(key: String): ArrayList<T>? = when {
    Version.isTiramisu -> getParcelableArrayListExtra(key, T::class.java)
    else -> @Suppress("deprecation") getParcelableArrayListExtra(key)
}

inline fun <reified T : Serializable> Intent.serializable(key: String): T? = when {
    Version.isTiramisu -> getSerializableExtra(key, T::class.java)
    else -> @Suppress("deprecation") getSerializableExtra(key) as? T
}

class TagMo : Application() {

    fun setThemePreference() {
        when (Preferences(applicationContext).applicationTheme()) {
            0 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
            1 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            2 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        }
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        if (isWearable)
            appContext = ScaledContext(this).watch(2f)
        if (Version.isLollipop) Reflection.unseal(base)
    }

    override fun onCreate() {
        super.onCreate()
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)

        val uiModeManager: UiModeManager = getSystemService(UI_MODE_SERVICE) as UiModeManager
        isWearable = uiModeManager.currentModeType == Configuration.UI_MODE_TYPE_WATCH || BuildConfig.WEAR_OS

        if (isWearable)
            appContext.setTheme(R.style.AppTheme)
        else
            setThemePreference()
        Thread.setDefaultUncaughtExceptionHandler { _: Thread?, error: Throwable ->
            val exception = StringWriter().apply {
                error.printStackTrace(PrintWriter(this))
            }
            try {
                Debug.sendException(this, exception.toString())
            } catch (_: Exception) { }
            try {
                Debug.saveException(this, exception.toString())
            } catch (_: Exception) { }
            android.os.Process.killProcess(android.os.Process.myPid())
            exitProcess(-1)
        }
        CoroutineScope(Dispatchers.IO).launch {
            ChromeIntegration.hasCustomTabs = ChromeIntegration.bindCustomTabService(this@TagMo)
        }
    }

    init { appContext = this }

    companion object {
        @JvmStatic
        lateinit var appContext : Context
            private set

        var isWearable = false

        const val uiDelay = 50

        val downloadDir by lazy { Storage.getDownloadDir("TagMo") }

        private const val commitHash = "#${BuildConfig.COMMIT}"
        val versionLabel = "TagMo ${BuildConfig.VERSION_NAME} (${
            if (BuildConfig.GOOGLE_PLAY) "Play" else "GitHub"
        } ${
            when {
                isWearable -> "Wear OS"
                BuildConfig.BUILD_TYPE == "release" -> "Release"
                else -> "Debug"
            }
        }) $commitHash"
        private const val commitLink = ("<a href=https://github.com/HiddenRamblings/TagMo/commit/"
                + BuildConfig.COMMIT + ">" + commitHash + "</a>")

        val versionLabelLinked = HtmlCompat.fromHtml(
                versionLabel.removePrefix("TagMo ").replace(commitHash, commitLink),
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )

        var hasSubscription = false

        const val isUserInputEnabled = false
    }
}